linode/manager
# feat: [M3-7159] - Update Enable Object Storage modal for beta DC-specific pricing

## Description 📝
This PR updates the copy in the Enable Object Storage modal to account for DC-specific pricing changes in the beta period. OBJ will remain free in Jakarta and Sao Paulo while in beta. After the beta ends, customers will be charged in these regions. 

## Major Changes 🔄
- Fixes a bug where the Close button ('X') did not close the modal
- Updates EnableObjectStorage modal to adhere to our coding standards (removed default export, renamed props).
- Creates and uses constants and a OBJ price increase map to determine regions with beta pricing.
   - Adds TODOs where this will be updated in a future ticket when DC-specific pricing is applied.
- Updates unit test coverage for EnableObjectStorageModal.

When the `dcSpecificPricing` feature flag is **enabled**:
- Adds copy to the Enable Object Storage modal in the buckets flow to inform users of free beta period
- Updates the copy of the Enable Object Storage modal in the access key flow to have more general messaging about pricing that applies to ALL regions, since no region selection is made
- Updates the "Learn more" link to direct to the OBJ section of https://linode.com/pricing rather than the OBJ docs page (https://www.linode.com/docs/products/storage/object-storage/#pricing)
   - Why? To help indicate that OBJ, like pricing for other products impacted by DC-specific pricing, can expect similar price changes in Jakarta and Sao Paulo once beta ends

## Preview 📷

| Description  | This Branch   |
| ------- | ------- |
| Prod / `objDCSpecificPricing` flag OFF | ![Screenshot 2023-09-19 at 9 12 58 PM](https://github.com/linode/manager/assets/114685994/3ae3f29c-8a3d-41e5-a1fe-509e20214940) |
|`objDCSpecificPricing` flag on with bucket in `us_east`  | ![Screenshot 2023-09-19 at 9 15 41 PM](https://github.com/linode/manager/assets/114685994/f8c290ac-f8ea-4df3-94de-06d3df5d0b22) |
|`objDCSpecificPricing` flag on with bucket in `id_cgk`  | ![Screenshot 2023-09-19 at 9 17 28 PM](https://github.com/linode/manager/assets/114685994/efd701a5-4e1e-491f-b918-b74e40625649) |
|`objDCSpecificPricing` flag on with bucket in `br_gru`  | ![Screenshot 2023-09-19 at 9 18 22 PM](https://github.com/linode/manager/assets/114685994/35361d8b-9a1d-4f85-a764-db82b485c3e2) |
|`objDCSpecificPricing` flag on no region selection (access keys) | ![Screenshot 2023-09-19 at 9 19 22 PM](https://github.com/linode/manager/assets/114685994/ecbcce06-414b-427d-9c5e-01eb3899f675) |

## How to test 🧪
1. **How to setup test environment?**
- Check out this PR.
- Turn on the `dcSpecificPricing` feature flag.
- If Object Storage already enabled, go to /account/settings and cancel it
2. **How to verify changes?**
- With the feature flag on, confirm that the copy of the modal matches the copy in the screenshots above for each scenario. 
To mock clusters in Sao Paulo or Jakarta, update the `rest.get('*object-storage/clusters'` mocked GET request in `serverHandlers.ts` to include clusters in both regions. Then turn on the MSW.
```
  rest.get('*object-storage/clusters', (req, res, ctx) => {
    const jakartaCluster = objectStorageClusterFactory.build({
      id: `id-cgk-0` as any,
      region: 'id-cgk',
    });
    const saoPauloCluster = objectStorageClusterFactory.build({
      id: `br-gru-0` as any,
      region: 'br-gru',
    });
    const clusters = objectStorageClusterFactory.buildList(3);
    return res(
      ctx.json(makeResourcePage([jakartaCluster, saoPauloCluster, ...clusters]))
    );
  }),
  rest.get('*object-storage/keys', (req, res, ctx) => {
    return res(
      ctx.json(makeResourcePage(objectStorageKeyFactory.buildList(3)))
    );
  }),
```
- With the feature flag off, confirm that modal in the access key flow and bucket flow (all regions) matches prod.
- Confirm an unrelated bug fix: clicking the Close button now closes the modal rather than does nothing.
4. **How to run Unit or E2E tests?**
Verify unit test passes:
```
yarn test EnableObjectStorageModal
```
